File: Objective-C Version/GooeySlideMenuDemo/GooeySlideMenuDemo/Classes/GooeySlideMenu.m
//
//  GooeySlideMenu.m
//  GooeySlideMenuDemo
//
//  Created by Kitten Yang on 15/8/11.
//  Copyright (c) 2015年 Kitten Yang. All rights reserved.
//

#import "GooeySlideMenu.h"
#import "SlideMenuButton.h"

#define buttonSpace 30
#define menuBlankWidth 50

@interface GooeySlideMenu()

@property (nonatomic,strong) CADisplayLink *displayLink;
@property  NSInteger animationCount; // 动画的数量
@end

@implementation GooeySlideMenu{
    
    UIVisualEffectView *blurView;
    UIView *helperSideView;
    UIView *helperCenterView;
    UIWindow *keyWindow;
    BOOL triggered;
    CGFloat diff;
    UIColor *_menuColor;
    CGFloat menuButtonHeight;
}


-(id)initWithTitles:(NSArray *)titles{
    return [self initWithTitles:titles withButtonHeight:40.0f withMenuColor:[UIColor colorWithRed:0 green:0.722 blue:1 alpha:1] withBackBlurStyle:UIBlurEffectStyleDark];
}

-(id)initWithTitles:(NSArray *)titles withButtonHeight:(CGFloat)height withMenuColor:(UIColor *)menuColor withBackBlurStyle:(UIBlurEffectStyle)style{
    
    self = [super init];
    if (self) {
        keyWindow = [[UIApplication sharedApplication]keyWindow];
        
        blurView = [[UIVisualEffectView alloc]initWithEffect:[UIBlurEffect effectWithStyle:style]];
        blurView.frame = keyWindow.frame;
        blurView.alpha = 0.0f;
        
        helperSideView = [[UIView alloc]initWithFrame:CGRectMake(-40, 0, 40, 40)];
        helperSideView.backgroundColor = [UIColor redColor];
        helperSideView.hidden = YES;
        [keyWindow addSubview:helperSideView];
        
        helperCenterView = [[UIView alloc]initWithFrame:CGRectMake(-40, CGRectGetHeight(keyWindow.frame)/2 - 20, 40, 40)];
        helperCenterView.backgroundColor = [UIColor yellowColor];
        helperCenterView.hidden = YES;
        [keyWindow addSubview:helperCenterView];
        
        self.frame = CGRectMake(- keyWindow.frame.size.width/2 - menuBlankWidth, 0, keyWindow.frame.size.width/2+menuBlankWidth, keyWindow.frame.size.height);
        self.backgroundColor = [UIColor clearColor];
        [keyWindow insertSubview:self belowSubview:helperSideView];
        
        _menuColor = menuColor;
        menuButtonHeight = height;
        [self addButtons:titles];
    }
    return self;
}

-(void)addButtons:(NSArray *)titles{
    if (titles.count % 2 == 0) {
        NSInteger index_down = titles.count/2;
        NSInteger index_up = -1;
        for (NSInteger i = 0; i < titles.count; i++) {
            NSString *title = titles[i];
            SlideMenuButton *home_button = [[SlideMenuButton alloc]initWithTitle:title];
            if (i >= titles.count / 2) {
                index_up ++;
                home_button.center = CGPointMake(keyWindow.frame.size.width/4, keyWindow.frame.size.height/2 + menuButtonHeight*index_up + buttonSpace*index_up + buttonSpace/2 + menuButtonHeight/2);
            }else{
                index_down --;
                home_button.center = CGPointMake(keyWindow.frame.size.width/4, keyWindow.frame.size.height/2 - menuButtonHeight*index_down - buttonSpace*index_down - buttonSpace/2 - menuButtonHeight/2);
            }
            
            home_button.bounds = CGRectMake(0, 0, keyWindow.frame.size.width/2 - 20*2, menuButtonHeight);
            home_button.buttonColor = _menuColor;
            [self addSubview:home_button];
            
            __weak typeof(self) WeakSelf = self;
            home_button.buttonClickBlock = ^(){
                [WeakSelf tapToUntrigger];
                WeakSelf.menuClickBlock(i,title,titles.count);
            };
        }
        
    }else{
        NSInteger index = (titles.count - 1) /2 +1;
        for (NSInteger i = 0; i < titles.count; i++) {
            index --;
            NSString *title = titles[i];
            SlideMenuButton *home_button = [[SlideMenuButton alloc]initWithTitle:title];
            home_button.center = CGPointMake(keyWindow.frame.size.width/4, keyWindow.frame.size.height/2 - menuButtonHeight*index - 20*index);
            home_button.bounds = CGRectMake(0, 0, keyWindow.frame.size.width/2 - 20*2, menuButtonHeight);
            home_button.buttonColor = _menuColor;
            [self addSubview:home_button];
            
            __weak typeof(self) WeakSelf = self;
            home_button.buttonClickBlock = ^(){
                [WeakSelf tapToUntrigger];
                WeakSelf.menuClickBlock(i,title,titles.count);
            };
        }
    }
    
}

- (void)drawRect:(CGRect)rect {

    UIBezierPath *path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointMake(0, 0)];
    [path addLineToPoint:CGPointMake(self.frame.size.width-menuBlankWidth, 0)];
    [path addQuadCurveToPoint:CGPointMake(self.frame.size.width-menuBlankWidth, self.frame.size.height) controlPoint:CGPointMake(keyWindow.frame.size.width/2+diff, keyWindow.frame.size.height/2)];
    [path addLineToPoint:CGPointMake(0, self.frame.size.height)];
    [path closePath];
    
    CGContextRef context = UIGraphicsGetCurrentContext();
    CGContextAddPath(context, path.CGPath);
    [_menuColor set];
    CGContextFillPath(context);
}

-(void)trigger{
    if (!triggered) {
        [keyWindow insertSubview:blurView belowSubview:self];
        [UIView animateWithDuration:0.3 animations:^{
            self.frame = self.bounds;
        }];
        
        [self beforeAnimation];
        [UIView animateWithDuration:0.7 delay:0.0f usingSpringWithDamping:0.5f initialSpringVelocity:0.9f options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionAllowUserInteraction animations:^{
            helperSideView.center = CGPointMake(keyWindow.center.x, helperSideView.frame.size.height/2);
        } completion:^(BOOL finished) {
            [self finishAnimation];
        }];
    
        [UIView animateWithDuration:0.3 animations:^{
            blurView.alpha = 1.0f;
        }];
        
        [self beforeAnimation];
        [UIView animateWithDuration:0.7 delay:0.0f usingSpringWithDamping:0.8f initialSpringVelocity:2.0f options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionAllowUserInteraction animations:^{
            helperCenterView.center = keyWindow.center;
        } completion:^(BOOL finished) {
            if (finished) {
                UITapGestureRecognizer *tapGes = [[UITapGestureRecognizer alloc]initWithTarget:self action:@selector(tapToUntrigger)];
                [blurView addGestureRecognizer:tapGes];
                [self finishAnimation];
            }
        }];
        [self animateButtons];
        triggered = YES;
    }else{
        [self tapToUntrigger];
    }
}

-(void)animateButtons{
    for (NSInteger i = 0; i < self.subviews.count; i++) {
        
        UIView *menuButton = self.subviews[i];
        menuButton.transform = CGAffineTransformMakeTranslation(-90, 0);
        [UIView animateWithDuration:0.7 delay:i*(0.3/self.subviews.count) usingSpringWithDamping:0.6f initialSpringVelocity:0.0f options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionAllowUserInteraction animations:^{
            menuButton.transform =  CGAffineTransformIdentity;
        } completion:NULL];
    }
    
}

-(void)tapToUntrigger{
    
    [UIView animateWithDuration:0.3 animations:^{
        self.frame = CGRectMake(-keyWindow.frame.size.width/2-menuBlankWidth, 0, keyWindow.frame.size.width/2+menuBlankWidth, keyWindow.frame.size.height);
    }];
    
    [self beforeAnimation];
    [UIView animateWithDuration:0.7 delay:0.0f usingSpringWithDamping:0.6f initialSpringVelocity:0.9f options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionAllowUserInteraction animations:^{
        helperSideView.center = CGPointMake(-helperSideView.frame.size.height/2, helperSideView.frame.size.height/2);
    } completion:^(BOOL finished) {
        [self finishAnimation];
    }];
    
    [UIView animateWithDuration:0.3 animations:^{
        blurView.alpha = 0.0f;
    }];
    
    [self beforeAnimation];
    [UIView animateWithDuration:0.7 delay:0.0f usingSpringWithDamping:0.7f initialSpringVelocity:2.0f options:UIViewAnimationOptionBeginFromCurrentState | UIViewAnimationOptionAllowUserInteraction animations:^{
        helperCenterView.center = CGPointMake(-helperSideView.frame.size.height/2, CGRectGetHeight(keyWindow.frame)/2);
    } completion:^(BOOL finished) {
        [self finishAnimation];
    }];
    
    triggered = NO;
    
}

//动画之前调用
-(void)beforeAnimation{
    if (self.displayLink == nil) {
        self.displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(displayLinkAction:)];
        [self.displayLink addToRunLoop:[NSRunLoop mainRunLoop] forMode:NSDefaultRunLoopMode];
    }
    self.animationCount ++;
}

//动画完成之后调用
-(void)finishAnimation{
    self.animationCount --;
    if (self.animationCount == 0) {
        [self.displayLink invalidate];
        self.displayLink = nil;
    }
}

-(void)displayLinkAction:(CADisplayLink *)dis{
    
    CALayer *sideHelperPresentationLayer   =  (CALayer *)[helperSideView.layer presentationLayer];
    CALayer *centerHelperPresentationLayer =  (CALayer *)[helperCenterView.layer presentationLayer];

    CGRect centerRect = [[centerHelperPresentationLayer valueForKeyPath:@"frame"]CGRectValue];
    CGRect sideRect = [[sideHelperPresentationLayer valueForKeyPath:@"frame"]CGRectValue];
        
    diff = sideRect.origin.x - centerRect.origin.x;
    
    [self setNeedsDisplay];
    
}

@end

File: Swift Version/GooeySlideMenuDemo-Swift/GooeySlideMenuDemo-Swift/Classes/GooeySlideMenu.swift
//
//  GooeySlideMenu.swift
//  GooeySlideMenuDemo-Swift
//
//  Created by Kitten Yang on 1/3/16.
//  Copyright © 2016 Kitten Yang. All rights reserved.
//

import UIKit

typealias MenuButtonClickedBlock = (index: Int, title: String, titleCounts: Int)->()

struct MenuOptions {
    var titles: [String]
    var buttonHeight: CGFloat
    var menuColor: UIColor
    var blurStyle: UIBlurEffectStyle
    var buttonSpace: CGFloat
    var menuBlankWidth: CGFloat
    var menuClickBlock: MenuButtonClickedBlock
}

class GooeySlideMenu: UIView {

    private var _option: MenuOptions
    private var keyWindow: UIWindow?
    private var blurView: UIVisualEffectView!
    private var helperSideView: UIView!
    private var helperCenterView: UIView!
    
    private var diff: CGFloat = 0.0
    private var triggered: Bool = false
    private var displayLink: CADisplayLink?
    private var animationCount: Int = 0
    
    init(options: MenuOptions) {
        _option = options
        if let kWindow = UIApplication.sharedApplication().keyWindow{
            keyWindow = kWindow
            let frame = CGRect(
                x: -kWindow.frame.size.width/2 - options.menuBlankWidth,
                y: 0,
                width: kWindow.frame.size.width/2 + options.menuBlankWidth,
                height: kWindow.frame.size.height)
            super.init(frame:frame)
        } else {
            super.init(frame:CGRectZero)
        }
        setUpViews()
    }

    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func drawRect(rect: CGRect) {
        let path = UIBezierPath()
        path.moveToPoint(CGPoint(x: 0, y: 0))
        path.addLineToPoint(CGPoint(x: frame.width-_option.menuBlankWidth, y: 0))
        path.addQuadCurveToPoint(CGPoint(x: frame.width-_option.menuBlankWidth, y: frame.height), controlPoint: CGPoint(x: frame.width-_option.menuBlankWidth+diff, y: frame.height/2))
        path.addLineToPoint(CGPoint(x: 0, y: frame.height))
        path.closePath()
        
        let context = UIGraphicsGetCurrentContext()
        CGContextAddPath(context, path.CGPath)
        _option.menuColor.set()
        CGContextFillPath(context)
    }
    
    func trigger() {
        if !triggered {
            if let keyWindow = keyWindow {
                keyWindow.insertSubview(blurView, belowSubview: self)
                UIView.animateWithDuration(0.3, animations: { [weak self] () -> Void in
                    self?.frame = CGRect(
                        x: 0,
                        y: 0,
                        width: keyWindow.frame.size.width/2 + _option.menuBlankWidth,
                        height: keyWindow.frame.size.height)
                })
                
                beforeAnimation()
                UIView.animateWithDuration(0.7, delay: 0.0, usingSpringWithDamping: 0.5, initialSpringVelocity: 0.9, options: [.BeginFromCurrentState,.AllowUserInteraction], animations: { [weak self] () -> Void in
                    self?.helperSideView.center = CGPointMake(keyWindow.center.x, helperSideView.frame.size.height/2);
                    }, completion: { [weak self] (finish) -> Void in
                        self?.finishAnimation()
                    })
                
                UIView.animateWithDuration(0.3, animations: { [weak self] () -> Void in
                    self?.blurView.alpha = 1.0
                })
                
                beforeAnimation()
                UIView.animateWithDuration(0.7, delay: 0.0, usingSpringWithDamping: 0.8, initialSpringVelocity: 2.0, options: [.BeginFromCurrentState,.AllowUserInteraction], animations: { [weak self] () -> Void in
                    self?.helperCenterView.center = keyWindow.center
                    }, completion: { [weak self] (finished) -> Void in
                        if finished {
                            let tapGesture = UITapGestureRecognizer(target: self, action: "tapToUntrigger")
                            self?.blurView.addGestureRecognizer(tapGesture)
                            self?.finishAnimation()
                        }
                })
                animateButtons()
                triggered = true
            }
        } else {
            tapToUntrigger()
        }
    }
    
}

extension GooeySlideMenu {
    
    private func setUpViews() {
        if let keyWindow = keyWindow {
            blurView = UIVisualEffectView(effect: UIBlurEffect(style: _option.blurStyle))
            blurView.frame = keyWindow.frame
            blurView.alpha = 0.0
            
            helperSideView = UIView(frame: CGRect(x: -40, y: 0, width: 40, height: 40))
            helperSideView.backgroundColor = UIColor.redColor()
            helperSideView.hidden = true
            keyWindow.addSubview(helperSideView)
            
            helperCenterView = UIView(frame: CGRect(x: -40, y: CGRectGetHeight(keyWindow.frame)/2 - 20, width: 40, height: 40))
            helperCenterView.backgroundColor = UIColor.yellowColor()
            helperCenterView.hidden = true
            keyWindow.addSubview(helperCenterView)
            
            backgroundColor = UIColor.clearColor()
            keyWindow.insertSubview(self, belowSubview: helperSideView)
            addButton()
        }
    }
    
    private func addButton() {
        let titles = _option.titles
        if titles.count % 2 == 0 {
            var index_down = titles.count / 2
            var index_up = -1
            for i in 0..<titles.count {
                let title = titles[i]
                let buttonOption = MenuButtonOptions(title:title, buttonColor:_option.menuColor, buttonClickBlock:{ [weak self] () -> () in
                    self?.tapToUntrigger()
                    self?._option.menuClickBlock(index: i,title: title,titleCounts: titles.count)
                    })
                let home_button = SlideMenuButton(option: buttonOption)
                home_button.bounds = CGRectMake(0, 0, frame.width - _option.menuBlankWidth - 20*2, _option.buttonHeight);
                addSubview(home_button)
                if (i >= titles.count / 2) {
                    index_up++
                    let y = frame.height/2 + _option.buttonHeight*CGFloat(index_up) + _option.buttonSpace*CGFloat(index_up)
                    home_button.center = CGPoint(x: (frame.width - _option.menuBlankWidth)/2, y: y+_option.buttonSpace/2 + _option.buttonHeight/2)
                } else {
                    index_down--
                    let y = frame.height/2 - _option.buttonHeight*CGFloat(index_down) - _option.buttonSpace*CGFloat(index_down)
                    home_button.center = CGPoint(x: (frame.width - _option.menuBlankWidth)/2, y: y - _option.buttonSpace/2 - _option.buttonHeight/2)
                }
            }
        } else {
            var index = (titles.count-1) / 2 + 1
            for i in 0..<titles.count {
                index--
                let title = titles[i]
                let buttonOption = MenuButtonOptions(title: title, buttonColor: _option.menuColor, buttonClickBlock: { [weak self] () -> () in
                    self?.tapToUntrigger()
                    self?._option.menuClickBlock(index: i, title: title, titleCounts: titles.count)
                    })
                let home_button = SlideMenuButton(option: buttonOption)
                home_button.bounds = CGRect(x: 0, y: 0, width: frame.width - _option.menuBlankWidth - 20*2, height: _option.buttonHeight)
                home_button.center = CGPoint(x: (frame.width - _option.menuBlankWidth)/2, y: frame.height/2 - _option.buttonHeight*CGFloat(index) - 20*CGFloat(index))
                addSubview(home_button)
            }
        }
    }
    
    private func animateButtons() {
        for i in 0..<subviews.count {
            let menuButton = subviews[i]
            menuButton.transform = CGAffineTransformMakeTranslation(-90, 0)
            UIView.animateWithDuration(0.7, delay: Double(i)*(0.3/Double(subviews.count)), usingSpringWithDamping: 0.6, initialSpringVelocity: 0.0, options: [.BeginFromCurrentState,.AllowUserInteraction], animations: { () -> Void in
                    menuButton.transform =  CGAffineTransformIdentity
                }, completion: nil)
        }
    }
    
   @objc private func tapToUntrigger() {
        UIView.animateWithDuration(0.3) { [weak self] () -> Void in
            self?.frame = CGRect(
                x: -keyWindow!.frame.size.width/2 - _option.menuBlankWidth,
                y: 0,
                width: keyWindow!.frame.size.width/2 + _option.menuBlankWidth,
                height: keyWindow!.frame.size.height)
        }
    
        beforeAnimation()
        UIView.animateWithDuration(0.7, delay: 0.0, usingSpringWithDamping: 0.6, initialSpringVelocity: 0.9, options: [.BeginFromCurrentState,.AllowUserInteraction], animations: { () -> Void in
            self.helperSideView.center = CGPoint(x: -self.helperSideView.frame.height/2, y: self.helperSideView.frame.height/2)
            }) { [weak self] (finish) -> Void in
                self?.finishAnimation()
        }
        
        UIView.animateWithDuration(0.3) { () -> Void in
            self.blurView.alpha = 0.0
        }
        
        beforeAnimation()
        UIView.animateWithDuration(0.7, delay: 0.0, usingSpringWithDamping: 0.7, initialSpringVelocity: 2.0, options: [.BeginFromCurrentState,.AllowUserInteraction], animations: { () -> Void in
            self.helperCenterView.center = CGPointMake(-self.helperSideView.frame.size.height/2, CGRectGetHeight(self.frame)/2)
            }) { (finish) -> Void in
                self.finishAnimation()
        }
        triggered = false
    }
    
    private func beforeAnimation() {
        if displayLink == nil {
            displayLink = CADisplayLink(target: self, selector: "handleDisplayLinkAction:")
            displayLink?.addToRunLoop(NSRunLoop.mainRunLoop(), forMode: NSDefaultRunLoopMode)
        }
        animationCount++
    }
    
    private func finishAnimation() {
        animationCount--
        if animationCount == 0 {
            displayLink?.invalidate()
            displayLink = nil
        }
    }
    
    @objc private func handleDisplayLinkAction(displaylink: CADisplayLink) {
        let sideHelperPresentationLayer = helperSideView.layer.presentationLayer() as! CALayer
        let centerHelperPresentationLayer = helperCenterView.layer.presentationLayer() as! CALayer
        
        let centerRect = centerHelperPresentationLayer.valueForKeyPath("frame")?.CGRectValue
        let sideRect   = sideHelperPresentationLayer.valueForKeyPath("frame")?.CGRectValue
        
        if let centerRect = centerRect, sideRect = sideRect {
            diff = sideRect.origin.x - centerRect.origin.x
        }
        setNeedsDisplay()
    }
}


